### PR TITLE
Faster startup of large distributed systems with latency

### PR DESCRIPTION
**Pull request by klauspost**
## Description

The current locking system has a fundamental problem for resolving high contention locks. This is only really expected for the initial startup lock, since servers try to get the same lock at the same time.

This is observed with high latency or large server count.

### Why it is hard to get `/transaction.lock`

1. All servers restart at the same time.
2. All servers sends a request to get `/transaction.lock` to all servers in a single set (up to 16).
3. When a server receives a request it will hand it out to the first one it sees (basically random).
4. Since each server hands out the lock to what is a random server no individual server can reach quorum.
5. All servers that got a lock releases them.
6. Each server waits 0 -> 50ms, then retries getting the lock. With enough servers and/or enough latency this is not enough to separate them.
7. Go to 3 (unless one server is lucky).

## Changes Explained

First of all we disable the automatic retries of lock acquirement. This is to strictly control when retries happens.

We insert a random spread of 1 second before we try to acquire the lock on startup.

When we cannot acquire a lock, wait 0 -> 2 seconds before we retry. This will make acquisition very likely to succeed within 3 seconds for most systems, with most taking less than 1 second.

We increase default lock retry time from 0->50ms to 0->100ms to still make it reasonably responsive and reduce the chance of overlap.

## How to test this PR?

A setup similar to above is required.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#edfe0dc2-a9c5-4c4e-af2e-40bf1228326b) |
| mint-dist-erasure.sh | :x: [more...](#4cfecc12-5e74-435d-b20c-edc7ef12f97e) |
<a name="4cfecc12-5e74-435d-b20c-edc7ef12f97e"></a>
### 16259-a1ead17/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.3:31204
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 0f60ab76d724:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 45 seconds
(9/14) Running minio-java tests ... FAILED in 25 seconds
{
  "name": "minio-java",
  "function": "getObjectRetention()",
  "duration": 49,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2022-12-17T18:02:29.2Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioAsyncClient.lambda$getObjectRetention$35(MinioAsyncClient.java:1738), java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642), java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506), java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073), io.minio.S3Base$1.onResponse(S3Base.java:559), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 49 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 40 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
<a name="edfe0dc2-a9c5-4c4e-af2e-40bf1228326b"></a>
### 16259-a1ead17/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.4:32625
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 9a80fe62f5c1:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 6 seconds
(5/14) Running awscli tests ... done in 57 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 12 seconds
(8/14) Running minio-go tests ... done in 31 seconds
(9/14) Running minio-java tests ... FAILED in 13 seconds
{
  "name": "minio-java",
  "function": "getObjectRetention()",
  "duration": 30,
  "status": "FAIL",
  "error": "java.time.format.DateTimeParseException: Text '2022-12-15T18:01:55.3Z' could not be parsed at index 19 >>> [io.minio.Xml.unmarshal(Xml.java:55), io.minio.MinioAsyncClient.lambda$getObjectRetention$35(MinioAsyncClient.java:1738), java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642), java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506), java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073), io.minio.S3Base$1.onResponse(S3Base.java:559), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 42 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 4 seconds
(11/14) Running s3cmd tests ... done in 12 seconds
(12/14) Running s3select tests ... done in 2 seconds
(13/14) Running versioning tests ... done in 3 minutes and 1 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

File: internal/dsync/drwmutex.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ const (
 	drwMutexRefreshInterval = 10 * time.Second
 
 	// maximum time to sleep before retrying a failed blocking lock()
-	lockRetryInterval = 50 * time.Millisecond
+	lockRetryInterval = 100 * time.Millisecond
```
Comment (harshavardhana) [inline internal/dsync/drwmutex.go]: Let's make this configurable? So that we can test a few values?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by vadmeste** [comment]
@klauspost maybe the random sleep can depend on the number of nodes as well, so it won't bottleneck when we have too many nodes. 

We have for now, globalEndpoints.Hostnames() which has the number of nodes

File: internal/dsync/drwmutex.go (hunk 1 above)
Comment (klauspost) [reply internal/dsync/drwmutex.go]: Sure. I'll merge patch by @vadmeste into this.

**Comment by klauspost** [comment]
@vadmeste The value doesn't depend too much on the number of hosts, but more on the contention of the lock and the latency variation.

We handle the high contention lock separately.

**Review by vadmeste — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**